mangaki/mangaki/utils/dpp.py

```diff
@@ -0,0 +1,139 @@
+        return 1 - squareform(pdist(self.matrix.T, metric=metric))
+
+
+class MangakiUniform():
+
+    def __init__(self, items):
+        self.items = items
+
+    def sample_k(self, nb_points):
+        return choice(self.items, nb_points).tolist()
+
+
+class MangakiDPP():
+
+    def __init__(self, items, similarity_matrix):
+        self.items = items
+        self.similarity_matrix = similarity_matrix
+
+    def sample_k(self, *args, **kwargs):
+        MAX_ITER = 10
+        for i in range(MAX_ITER):
+            try:
+                return self._sample_k(*args, **kwargs)
+            except np.linalg.linalg.LinAlgError as e:
+                print('LinAlgError in MangakiDPP')
+        raise ValueError('Too much LinAlgError')
+
+    def _sample_k(self, k, max_nb_iterations=1000, rng=np.random):
+        """
+        Thanks to mehdidc on github : https://github.com/mehdidc/dpp
+        Sample a list of k items from a DPP defined
+        by the similarity matrix L. The algorithm
+        is iterative and runs for max_nb_iterations.
+        The algorithm used is from
+        (Fast Determinantal Point Process Sampling withw
+        Application to Clustering, Byungkon Kang, NIPS 2013)
+        """
+        items = self.items
+        L = self.similarity_matrix
+        initial = rng.choice(range(len(items)), size=k, replace=False)
+        X = [False] * len(items)
+        for i in initial:
+            X[i] = True
+        X = np.array(X)
+        for i in range(max_nb_iterations):
+            u = rng.choice(np.arange(len(items))[X])
+            v = rng.choice(np.arange(len(items))[~X])
+            Y = X.copy()
+            Y[u] = False
+            L_Y = L[Y, :]
+            L_Y = L_Y[:, Y]
+            L_Y_inv = np.linalg.inv(L_Y)
+            c_v = L[v:v + 1, :]
+            c_v = c_v[:, v:v + 1]
+            b_v = L[Y, :]
+            b_v = b_v[:, v:v + 1]
+            c_u = L[u:u + 1, :]
+            c_u = c_u[:, u:u + 1]
+            b_u = L[Y, :]
+            b_u = b_u[:, u:u + 1]
+            p = min(1, c_v - np.dot(np.dot(b_v.T, L_Y_inv), b_v) /
+                    (c_u - np.dot(np.dot(b_u.T, L_Y_inv.T), b_u)))
+            if rng.uniform() <= p:
+                X = Y[:]
+                X[v] = True
+        return np.array(items)[X]
+
+
+def compare(similarity, algos, nb_points, nb_iterations=20):
+
+    resultats = np.zeros([len(algos), 2])
+
+    for _ in range(nb_iterations):
+        for i in range(len(algos)):
+            items = algos[i].sample_k(nb_points)
+            points = similarity.matrix[:, items].T.toarray()
+
+            det = np.linalg.det(squareform(pdist(
+                points,
+                metric='cosine')))
+
+            diam = diameter_0(points)
+
+            resultats[i, 0] += det
+            resultats[i, 1] += diam
+    resultats /= nb_iterations
+
+    return resultats
+
+
+if __name__ == '__main__':
+    build_matrix = RatingsMatrix(fname='/home/voisin/mangaki/data/ratings.csv')
```

Comment:
OK, j'ai misread sur les dict comprehensions ces histoires d'inversion !
Entendu @ stage de François. J'aimerais juste qu'on ouvre une issue pour
suivre les choses à faire par dessus?

On Tue, Aug 2, 2016, 19:47 Basile Clement <notifications@github.com> wrote:

> In mangaki/mangaki/utils/dpp.py
> <https://github.com/mangaki/mangaki/pull/201#discussion_r73133480>:
>
> > +
> > +            det = np.linalg.det(squareform(pdist(
> > +                points,
> > +                metric='cosine')))
> > +
> > +            diam = diameter_0(points)
> > +
> > +            resultats[i, 0] += det
> > +            resultats[i, 1] += diam
> > +    resultats /= nb_iterations
> > +
> > +    return resultats
> > +
> > +
> > +if __name__ == '__main__':
> > +    build_matrix = RatingsMatrix(fname='/home/voisin/mangaki/data/ratings.csv')
>
> En fait, cette partie dans le if __name__ == '__main__' devrait être
> entièrement supprimée.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mangaki/mangaki/pull/201/files/fbb01b1c9d60d3b5042f0dbc1417171c50c45e41#r73133480>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATMxOT9UR5KIL1VUIPqE61b9_Fdi35dks5qbyAkgaJpZM4JEHTk>
> .
>
-- 

-- 

Ryan Lahfa, from my phone.